### ISSUE TITLE
Provide an analyzer to check if an argument name matches the name of another parameter in the method

### ISSUE THREAD

**Issue by tannergooding** [dotnet/roslyn-analyzers#1216]
For example, imagine you have

```C#
// Defines a RGBA Color
public struct Color
{
    public Color(float red, float green, float blue, float alpha);
}
```

Then, you create a new Color `var color = new Color(red, green, blue, alpha)`. In this case, everything is good, the names of our arguments match up with the name of the parameter.

Now, imagine that instead you have:

```C#
// Defines a BGRA Color
public struct Color
{
    public Color(float blue, float green, float red, float alpha);
}
```

Then, you create a new Color `var color = new Color(red, green, blue, alpha)`. Such an analyzer would then warn you that you passed an argument 'red' to the parameter 'blue', but there is a parameter 'red' at a different index that you may have meant instead.

**Comment by tannergooding** [comment]
This would also be useful to catch cases like `new ArgumentException()` where parameters are ordered `(string message, string paramName)`, which differs from things like `ArgumentNullException` and `ArgumentOutOfRangeException`, whose parameters are ordered `(string paramName, string message)`.

**Comment by mavasani** [comment]
I think a better analyzer here would enforce used of named arguments when you have successive parameters of the same type (or convertible to same type with implicit conversions).

**Comment by mavasani** [comment]
Tagging @Pilchie @sharwell @CyrusNajmabadi - I think this is a good candidate for an IDE Code style analyzer. Thoughts?

**Comment by tannergooding** [comment]
I think both would be useful. There are definitely users who would want both, especially if we take this relaxation on named arguments: https://github.com/dotnet/csharplang/issues/518

**Comment by 333fred** [comment]
@mavasani definitely think both would be a good idea. What if I have a function that takes a couple of `int`s as the first two parameters, and then a couple of unrelated parameters? I'd hate to be forced to use named parameters for all of them.

**Comment by CyrusNajmabadi** [comment]
This is not really code style.  I think it would be fine as a general analyzer produced by the roslyn-analyzers package.

**Comment by sharwell** [comment]
I would leverage the rules for inferring tuple names here - if the effective name of an argument value matches a different parameter name, report a diagnostic.

**Comment by mykolav** [comment]
Hello,

Here's such an analyzer implemented as a proof-of-concept - https://github.com/mykolav/use-named-arguments
And the same analyzer implemented in `F#` - https://github.com/mykolav/require-named-args-fs

At the moment only `C#` is supported.   
Ideally, I'd like to add `VB` support and make a pull request to this repo. But given my other commitments, I'm not sure I can prioritize it to happen in a reasonable time-frame.

If anyone is willing to fork and make a pull request, I'll be glad to answer any questions about the code.

**Comment by Evangelink** [comment]
Is it ok to start working on this rule? If so, I just wanted to clarify one use-case:

```csharp
// Defines a BGRA Color
public struct Color
{
    public Color(float blue, float green, float red, float alpha);
}

var new Color(blue: red,...); // raise or not?
```

On the case described before, are you ok with not raising the diagnostic?

**Comment by danpetitt** [comment]
Love to see this in Visual Studio natively with ability to apply/fix an entire solution

**Comment by sharwell** [comment]
The main problem I see with this rule is variables rarely exactly match the names of parameters (based on statistical observations from the argument inference testing). I wonder if parameter name hints might be a better solution for this. Perhaps an option to automatically show parameter name hints if two or more arguments of an invocation are names that do not match the parameter names and the arguments are interchangeable based on their types?
